File: kpis/aeb/kpiDistance.m
function kpiDistance(obj, i, aebStartIdx, aebEndIdx)
    % Determine if PB and FB are activated during AEB event
    % Assuming PB is indicated by aebTargetDecel dropping below -6 and FB below -12
    %% Inputs:
    % obj            - KPIExtractor object
    % i              - Index of the file in kpiTable
    % aebStartIdx    - Index of AEB request event
    % aebEndIdx      - Index of AEB end event
    %% Outputs:
    % kpiTable       - Updated table with distance KPIs (for consistency)   

    % Bind signal data and longGap
    longGap = obj.signalMatChunk.longGap;

    % Ensure required columns exist in obj.kpiTable
    if ~ismember('firstDetDist', obj.kpiTable.Properties.VariableNames)
        obj.kpiTable.firstDetDist = NaN(height(obj.kpiTable), 1);    
    end
    if ~ismember('stableDetDist', obj.kpiTable.Properties.VariableNames)
        obj.kpiTable.stableDetDist = NaN(height(obj.kpiTable), 1);
    end
    if ~ismember('aebIntvDist', obj.kpiTable.Properties.VariableNames)
        obj.kpiTable.aebIntvDist = NaN(height(obj.kpiTable), 1);
    end
    if ~ismember('aebStopGap', obj.kpiTable.Properties.VariableNames)
        obj.kpiTable.aebStopGap = NaN(height(obj.kpiTable), 1);
    end

    % Consider the segment up to end of AEB event
    segment = longGap(1:aebEndIdx);

    % Calculate firstDetDist: First non-zero distance
    firstNonZeroIdx = find(segment ~= 0, 1, 'first');
    if ~isempty(firstNonZeroIdx)
        obj.kpiTable.firstDetDist(i) = longGap(firstNonZeroIdx);
    else
        obj.kpiTable.firstDetDist(i) = NaN;
    end

    % Calculate stableDetDist: Distance at the satrt of the last continuous non-zero segment before aebEndIdx
    lastNonZeroIdx = find(segment ~= 0, 1, 'last');
    if ~isempty(lastNonZeroIdx)
        trimmed = segment(1:lastNonZeroIdx);
        lastZeroIdx = find(trimmed == 0, 1, 'last');
        if isempty(lastZeroIdx)
            stableIdx = firstNonZeroIdx;
        else
            stableIdx = lastZeroIdx + 1;
        end
        obj.kpiTable.stableDetDist(i) = segment(stableIdx);
    else
        obj.kpiTable.stableDetDist(i) = NaN;
    end

    obj.kpiTable.aebIntvDist(i) = longGap(aebStartIdx);
    obj.kpiTable.aebStopGap(i) = longGap(aebEndIdx);
end

File: python/kpis/aeb/kpi_distance.py
import numpy as np
import pandas as pd

def kpi_distance(obj, i, aeb_start_idx, aeb_end_idx):
    """
    Compute distance KPIs during AEB event.
    Updates obj.kpi_table in place.

    Parameters
    ----------
    obj : object with attributes
        - kpi_table : pandas.DataFrame
        - signal_mat_chunk : dict-like with array "longGap"
    i : int
        Index of the file in kpi_table.
    aeb_start_idx : int
        Index of AEB request event.
    aeb_end_idx : int
        Index of AEB end event.
    """
    kpi_table = obj.kpi_table
    long_gap = np.asarray(obj.signal_mat_chunk["longGap"])

    # Ensure required columns exist
    for col in ["firstDetDist", "stableDetDist", "aebIntvDist", "aebStopGap"]:
        if col not in kpi_table.columns:
            kpi_table[col] = pd.Series([np.nan] * len(kpi_table), dtype="float")

    # Consider segment up to AEB end
    segment = long_gap[: aeb_end_idx + 1]

    # First detection distance (first non-zero)
    nonzero_idx = np.flatnonzero(segment != 0)
    if len(nonzero_idx) > 0:
        first_nonzero_idx = nonzero_idx[0]
        kpi_table.at[i, "firstDetDist"] = long_gap[first_nonzero_idx]
    else:
        kpi_table.at[i, "firstDetDist"] = np.nan

    # Stable detection distance (start of last continuous non-zero segment)
    if len(nonzero_idx) > 0:
        last_nonzero_idx = nonzero_idx[-1]
        trimmed = segment[: last_nonzero_idx + 1]
        zero_idx = np.flatnonzero(trimmed == 0)
        if len(zero_idx) == 0:
            stable_idx = first_nonzero_idx
        else:
            stable_idx = zero_idx[-1] + 1
        kpi_table.at[i, "stableDetDist"] = segment[stable_idx]
    else:
        kpi_table.at[i, "stableDetDist"] = np.nan

    # AEB intervention distance at start and end
    kpi_table.at[i, "aebIntvDist"] = long_gap[aeb_start_idx]
    kpi_table.at[i, "aebStopGap"] = long_gap[aeb_end_idx]
